feat: adding collaborator via invitation link

vedant-jain03 opened the pull request:
Fixes #2307

#### Describe the changes you have made in this PR -
- Migration file to add new column collaboration_token, index, and token_expires_at in Project table
- Functionality to add members to the group via link
- Different conditions like if the collaborator is already present, URL not valid, token expired.
- Expiry date of token is 12 days.

### Screenshots of the changes (If any) -

https://user-images.githubusercontent.com/76901313/144847443-cc12d46e-7887-4f3d-8ba8-9cbc4bc81415.mp4




Note: Please check **Allow edits from maintainers.** if you would like us to assist in the PR.

vedant-jain03 committed:
feat: adding collaborator via invitation link

vedant-jain03 committed:
fixes: typo

vedant-jain03 commented:
@tachyons @pavanjoshi914 Kindly Review.

tachyons commented:
You can follow the rest conventions by tweaking the URL. 

Token generation should be 

POST /projects/:id/collaboration

vedant-jain03 committed:
fixes: changes the request of collaboration_token from put to post

vedant-jain03 commented:
@tachyons Kindly review

vedant-jain03 committed:
fixes: codeclimate issue

vedant-jain03 commented:
@tachyons I have a suggestion, what is we can also convert the design of showing collaborators just like showing group members. As it is space optimized.

nitin10s commented:
Let's leave it as it is, for now, we'll revisit certain sections like these and design them appropriately.

vedant-jain03 commented:
> Let's leave it as it is, for now, we'll revisit certain sections like these and design them appropriately.

Sure

gr455 commented on app/controllers/projects_controller.rb:
```diff
@@ -36,6 +36,22 @@ def show
 
+  def project_invite
+    if Project.with_project_token.exists?(collaboration_token: params[:token])
+      if current_user.collaborations.exists?(project: @project)
```
When user is unauthenticated, current_user is nil. So better use `current_user&.collaborations....`

gr455 reviewed (changes requested):
This doesn't work when the Project access type is Private, although I can add collaborators in Private projects by email

vedant-jain03 commented:
> This doesn't work when the Project access type is Private, although I can add collaborators in Private projects by email

We can make this functionality disable for private project. WDYT?

vedant-jain03 committed:
fix: authenticated user results

gr455 commented:
> > This doesn't work when the Project access type is Private, although I can add collaborators in Private projects by email
> 
> I think this is working, have a look:
> 
>  bandicam.2022-01-06.13-09-34-943.mp4

Oh, I had pasted the link in an unauthenticated session (incognito). Instead of directing me to login, it just showed me the unauthorized message

vedant-jain03 commented:
> > > This doesn't work when the Project access type is Private, although I can add collaborators in Private projects by email
> > 
> > 
> > I think this is working, have a look:
> > bandicam.2022-01-06.13-09-34-943.mp4
> 
> Oh, I had pasted the link in an unauthenticated session (incognito). Instead of directing me to login, it just showed me the unauthorized message

Oh so WDYT?

vedant-jain03 commented:
@tachyons Kindly provide me feedback for this.

tachyons commented on app/controllers/projects_controller.rb:
```diff
@@ -36,6 +36,22 @@ def show
     commontator_thread_show(@project)
   end
 
+  def project_invite
```
move this to project/invitations controller

vedant-jain03 replied:
Sure

tachyons commented on app/views/projects/_add_collaborator_modal.html.erb:
```diff
@@ -31,13 +31,28 @@
+          <span><%= t("projects.invite_link_btn") %></span>
+        <% end %>
+        <div class="collapse" id="multiCollapseExample2">
+          <textarea id="embedInviteLink" class="collaborator-invite-link-text" rows="4" cols="35"></textarea>
```
@vedant-jain03 Why this is closed ?

vedant-jain03 replied:
@tachyons I thought it is just readonly textarea as we are just rendering the link for the invitation, there should not be any modification from the user end to the invitation link.
WDYT should I add label for it sir @tachyons ?

tachyons commented on app/views/projects/_add_collaborator_modal.html.erb:
```diff
@@ -31,13 +31,28 @@
 
 <script>
 $(document).ready(() => {
   // docs: http://bootstrap-tagsinput.github.io/bootstrap-tagsinput/examples/
```
This comment should be moved

tachyons commented on app/views/projects/_add_collaborator_modal.html.erb:
```diff
@@ -31,13 +31,28 @@
 $(document).ready(() => {
   // docs: http://bootstrap-tagsinput.github.io/bootstrap-tagsinput/examples/
+  $(".collaborator-invite-link-text").html("<%= user_project_url %>/invite/<%= @project.collaboration_token %>");
+  $('#invite_collab_button').on('click', ()=>{
```
Use stimulas js

vedant-jain03 committed:
fix: stimulus intergation, project/invitation controller, added readonly to textarea

vedant-jain03 committed:
fix: refactor create method

vedant-jain03 commented:
@tachyons 
- I have added integration with the stimulus controller.
- Move project_invite to project/invitation#create.
- Added check if the collaborator is the author.
- Added `readonly` property to textarea.

tachyons commented on app/javascript/controllers/projectCollaboration_controller.js:
```diff
@@ -0,0 +1,10 @@
+export default class extends Controller {
+    connect() {
+        $('#invite_collab_button').on('click', () => {
+            setTimeout(function () { $("#embedInviteLink").select(); document.execCommand("copy") }, 100);
```
Refer https://stimulus.hotwired.dev/reference/actions for onclick events in stimulas

vedant-jain03 committed:
fix: onclick event integration using stimulus

vedant-jain03 committed:
Update _add_collaborator_modal.html.erb

vedant-jain03 committed:
Merge branch 'master' into issue2307

vedant-jain03 commented:
@tachyons 
- Added data-action = click for embed Invitation link

vedant-jain03 committed:
Update projectCollaboration_controller.js

vedant-jain03 commented:
@tachyons Kindly review sir.

vedant-jain03 commented:
@gr455 Any views for this PR?

vedant-jain03 commented:
@tachyons Any feedback for this, sir?

tachyons reviewed (changes requested):
codeclimate warnings are closed without addressing them

aryanndwi123 commented:
I was going through the code and this was excellent work. Thank you, @vedant-jain03, for your contribution!

I'm closing this PR since the issue is already being worked on by @Yana-do-code in #7426.
